src/diffusers/loaders.py

```diff
@@ -3334,3 +3344,170 @@ def _remove_text_encoder_monkey_patch(self):
         else:
             self._remove_text_encoder_monkey_patch_classmethod(self.text_encoder)
             self._remove_text_encoder_monkey_patch_classmethod(self.text_encoder_2)
+
+
+class IPAdapterMixin:
+    """Mixin for handling IP Adapters."""
+
+    def set_ip_adapter(self):
+        unet = self.unet
+        attn_procs = {}
+        for name in unet.attn_processors.keys():
+            cross_attention_dim = None if name.endswith("attn1.processor") else unet.config.cross_attention_dim
+            if name.startswith("mid_block"):
+                hidden_size = unet.config.block_out_channels[-1]
+            elif name.startswith("up_blocks"):
+                block_id = int(name[len("up_blocks.")])
+                hidden_size = list(reversed(unet.config.block_out_channels))[block_id]
+            elif name.startswith("down_blocks"):
+                block_id = int(name[len("down_blocks.")])
+                hidden_size = unet.config.block_out_channels[block_id]
+            if cross_attention_dim is None:
+                attn_processor_class = (
+                    AttnProcessor2_0 if hasattr(F, "scaled_dot_product_attention") else AttnProcessor
+                )
+                attn_procs[name] = attn_processor_class()
+            else:
+                attn_processor_class = (
+                    IPAdapterAttnProcessor2_0 if hasattr(F, "scaled_dot_product_attention") else IPAdapterAttnProcessor
+                )
+                attn_procs[name] = attn_processor_class(
+                    hidden_size=hidden_size, cross_attention_dim=cross_attention_dim, scale=1.0
+                ).to(dtype=unet.dtype, device=unet.device)
+
+        unet.set_attn_processor(attn_procs)
+
+        if hasattr(self, "controlnet"):
+            attn_processor_class = (
+                IPAdapterControlNetAttnProcessor2_0
+                if hasattr(F, "scaled_dot_product_attention")
+                else IPAdapterControlNetAttnProcessor
+            )
+            self.pipeline.controlnet.set_attn_processor(attn_processor_class())
+
+    def load_ip_adapter(
```

Comment:
Can you explain the rationale?

Reply:
ohh sounds great! 

so we don't have to pass the `image_encoder` to `pipe.from_pretrained()` any more - we can just automatically pick it up from the ip-adapter weights folder here. we can add a optional path feature extractor?

Reply:
or maybe not needed 
they just use the default config for the CLIPProcessor

https://github.com/tencent-ailab/IP-Adapter/blob/e317f750e1829e846c944b9fdd7a2f2842c46501/ip_adapter/ip_adapter.py#L79

Reply:
> load CLIP Image encoder and feature extractor

I added code to do this! pretty happy about it. Now that if the image_encoder is not added to the pipeline yet and the weight folder come with a image encoder , it can automatically loaded here. But user still has the option to manually load it and pass it to the pipeline. I updated the doc to reflect this. 

I don't think user ever need to worry abut loading feature_extractor - if it is not already a components in pipeline, we can always add it here with default setting (that's what's used in original repo). So I removed the mention from the doc to keep it simple. 

> 3. load adaptable image proj and cross-attn weigths
       # 4. call unet._load_ip_adaptive_weights(....)

I created a `unet._load_ip_adaptive_weights(state_dict)` that does both 3 and 4 - #4851 let me know if this is something you have in mind. Happy to reiterate if it is not!

Reply:
The rationale here is simply to seperate concerns - the unet should be responsible for loading the trainable ip projection layers. It's not the pipeline's responsibility. The pipeline should only care about loading the image encoder and feature extractor and then offload the other responsability to the unet. Advantages:
- 1.) It's cleaner
- 2.) It's easier to test (just need to test the unet class now)
- 3.) It's easier to extend to new use cases